piker/ui/_search.py

```diff
@@ -0,0 +1,961 @@
+# piker: trading gear for hackers
+# Copyright (C) Tyler Goodlet (in stewardship for piker0)
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""
+qompleterz: embeddable search and complete using trio, Qt and fuzzywuzzy.
+
+"""
+
+# link set for hackzin on this stuff:
+# https://doc.qt.io/qt-5/qheaderview.html#moving-header-sections
+# https://doc.qt.io/qt-5/model-view-programming.html
+# https://doc.qt.io/qt-5/modelview.html
+# https://doc.qt.io/qt-5/qtreeview.html#selectedIndexes
+# https://doc.qt.io/qt-5/qmodelindex.html#siblingAtColumn
+# https://doc.qt.io/qt-5/qitemselectionmodel.html#currentIndex
+# https://www.programcreek.com/python/example/108109/PyQt5.QtWidgets.QTreeView
+# https://doc.qt.io/qt-5/qsyntaxhighlighter.html
+# https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/completion/completiondelegate.py#L243
+# https://forum.qt.io/topic/61343/highlight-matched-substrings-in-qstyleditemdelegate
+
+from collections import defaultdict
+from contextlib import asynccontextmanager
+from functools import partial
+from typing import (
+    List, Optional, Callable,
+    Awaitable, Sequence, Dict,
+    Any, AsyncIterator, Tuple,
+)
+import time
+# from pprint import pformat
+
+from fuzzywuzzy import process as fuzzy
+import trio
+from trio_typing import TaskStatus
+from PyQt5 import QtCore, QtGui
+from PyQt5 import QtWidgets
+from PyQt5.QtCore import (
+    Qt,
+    # QSize,
+    QModelIndex,
+    QItemSelectionModel,
+)
+from PyQt5.QtGui import (
+    # QLayout,
+    QStandardItem,
+    QStandardItemModel,
+)
+from PyQt5.QtWidgets import (
+    QWidget,
+    QTreeView,
+    # QListWidgetItem,
+    # QAbstractScrollArea,
+    QStyledItemDelegate,
+)
+
+
+from ..log import get_logger
+from ._style import (
+    _font,
+    DpiAwareFont,
+    # hcolor,
+)
+
+
+log = get_logger(__name__)
+
+
+class SimpleDelegate(QStyledItemDelegate):
```

Comment:
Can actually probably drop this for now but may end up being handy for syntax highlight.
See #186.

Reply:
Nah, gonna leave it in, can always trim later.